learning_observer/learning_observer/canvas.py

```diff
@@ -0,0 +1,341 @@
+import os
+import json
+import string
+import requests
+import recordclass
+import configparser
+import aiohttp
+import aiohttp.web
+import pmss
+import yaml
+
+import learning_observer.settings as settings
+import learning_observer.log_event
+import learning_observer.util
+import learning_observer.auth
+import learning_observer.runtime
+
+cache = None
+
+pmss.register_field(
+    name="default_server",
+    type=pmss.pmsstypes.TYPES.string,
+    description="The default server for Canva",
+    required=True
+)
+pmss.register_field(
+    name="client_id",
+    type=pmss.pmsstypes.TYPES.string,
+    description="The client ID for Canva",
+    required=True
+)
+pmss.register_field(
+    name="client_secret",
+    type=pmss.pmsstypes.TYPES.string,
+    description="The client secret for Canva",
+    required=True
+)
+pmss.register_field(
+    name="access_token",
+    type=pmss.pmsstypes.TYPES.string,
+    description="The access token for Canva",
+    required=True
+)
+pmss.register_field(
+    name="refresh_token",
+    type=pmss.pmsstypes.TYPES.string,
+    description="The refresh token for Canva",
+    required=True
+)
+
+class Endpoint(recordclass.make_dataclass("Endpoint", ["name", "remote_url", "doc", "cleaners"], defaults=["", None])):
+    def arguments(self):
+        return extract_parameters_from_format_string(self.remote_url)
+
+    def _local_url(self):
+        parameters = "}/{".join(self.arguments())
+        base_url = f"/canvas/{self.name}"
+        if len(parameters) == 0:
+            return base_url
+        else:
+            return base_url + "/{" + parameters + "}"
+
+    def _add_cleaner(self, name, cleaner):
+        if self.cleaners is None:
+            self.cleaners = dict()
+        self.cleaners[name] = cleaner
+        if 'local_url' not in cleaner:
+            cleaner['local_url'] = self._local_url + "/" + name
+
+    def _cleaners(self):
+        if self.cleaners is None:
+            return []
+        else:
+            return self.cleaners
+
+ENDPOINTS = list(map(lambda x: Endpoint(*x), [
+    ("course_list", "/courses"),
+    ("course_roster", "/courses/{courseId}/students"),
+    ("course_work", "/courses/{courseId}/assignments"),
+    ("coursework_submissions", "/courses/{courseId}/assignments/{assignmentId}/submissions"),
+]))
+
+def extract_parameters_from_format_string(format_string):
+    '''
+    Extracts parameters from a format string. E.g.
+
+    >>> ("hello {hi} my {bye}")]
+    ['hi', 'bye']
+    '''
+    # The parse returns a lot of context, which we discard. In particular, the
+    # last item is often about the suffix after the last parameter and may be
+    # `None`
+    return [f[1] for f in string.Formatter().parse(format_string) if f[1] is not None]
+
+class Canvas:
```

Comment:
I'm a little confused by this class and it's purpose. A docstring would be super helpful.

This class may not be needed entirely (see my note above about how the auth tokens ought to work within the system).

Reply:
My refactoring has done away with the class